contrib_bots/lib/giveaway.py

```diff
@@ -0,0 +1,215 @@
+
+    def usage(self):
+        return '''
+            This plugin will allow for users to create
+            giveaways which people can enter. After a set
+            period of time has passed, a random person will
+            be chosen from the list of random people whom
+            entered the giveaway.
+
+            Syntax:
+            @giveaway command giveaway_name giveaway_time
+
+            Command:
+                - init: initialize a giveaway
+                - enter: enter a giveaway (only works if a
+                giveaway is already running)
+                - exit: exit
+                - cancel
+
+            This plugin was created by Daniel O'Brien
+            as part of Google Code-In 2016-2017
+            '''
+
+    def triage_message(self, message, client):
+        # return True if we want to (possibly) response to this message
+
+        original_content = message['content']
+        if message['display_recipient'] == 'giveawayer':
+            return False
+        return original_content.startswith('@giveaway')
+
+    def handle_message(self, message, client, state_handler):
+
+        original_content = message['content']
+        original_sender = message['sender_email']
+        if original_content.split(" ")[1] == "init":
+            print("Giveaway: init")
+            if not self.giveaway_on:
+                print("Giveaway: No giveaway currently running")
+                try:
+                    self.giveaway_name = original_content.split(" ")[2]
+                    self.giveaway_creator = original_sender
+                    try:
+                        self.giveaway_timeMins = round(float(original_content.split(" ")[3]), 1)
+                        if self.giveaway_timeMins > 100:
+                            # Giveaway time must be less than 100 mins to avoid stack overflow.
+                            # However, limit can be increased within python with the command
+                            # sys.setrecursionlimit(NewLimit), default is 1000
+                            new_content = 'Error: Giveaway time must be less than or equal to 100 mins'
+                        else:
+                            self.giveaway_time = self.giveaway_timeMins * 60
+                            new_content = ("""{} has initiated a giveaway called {} for {} minutes"""
+                                           .format(message['sender_full_name'], self.giveaway_name, self.giveaway_timeMins))
+                            self.giveaway_on = True
+                            self.giveaway_start_time = time.time()
+                            self.giveawayStream = message['display_recipient']
+                            self.giveaway_subject = message['subject']
+                            self.giveaway_client = client
+                            self.giveaway_stage = 0
+                            self.one_minute_warning = False
+                            self.check_time()
+                    except:
```

Comment:
Better to handle specific exception and keep the size of `try` block to minimum.